### PR TITLE
FSDP: supporting gradient accumulation without no_sync context manager to save GPU memory

### PR DESCRIPTION
**Pull request by myleott**
Ideally we can do gradient accumulation without the `no_sync` context. In this case it should do the reduce-scatter in the backward, but accumulate into the sharded gradient so that memory usage is small.

### CODE REVIEW

**Comment by min-xu-ai** [comment]
Hi @myleott, I think this is ready. Do you want to give it a spin?

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -1278,7 +1284,8 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
             # Free full params. As a special case, we don't free the full params
             # when in a ``no_sync`` context (as inversely indicated by
             # ``self._require_backward_grad_sync``), since the params will not
-            # get updated before the next forward.
+            # get updated before the next forward. This saves networking
+            # bandwidth but uses more GPU memory.
             self._free_full_params([param])
```
Comment (zhaojuanmao) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1289]: I'm wondering whether full params could be freed and local grad shard is kept in 'no_sync' mode, in this case, 'no_sync' can save GPU memory, right?

Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1289]: Good question. In no sync mode, it is both full params and full grads that are consuming the GPU memory.

- We keep full params since the params are not updated while we accumulate the grads
- We keep the full grads since we are not reducing the scatter them (i.e. shard them)

In this new change, we free and reshard full params and shard grads and accumulate into grad shards. So both of the memory overhead above are avoided at the cost of networking overhead.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -783,19 +784,23 @@ def load_local_state_dict(
     @contextlib.contextmanager
     def no_sync(self) -> Generator:
```
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:785]: this naming might be canonical and this comment might be out of scope, but `no_grad_sync` would be a clearer name.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -1536,10 +1549,23 @@ def _use_full_params(self) -> None:
...
+                    pass
+                elif p.grad.size() == p._fp32_shard.shape:
+                    # This is gradient accumulation without no_sync context.
+                    # We save the grad shard and set p.grad to None for this backward pass.
+                    # We will accumulate after this pass's grad is generated and reduced and
+                    # sharded.
+                    p._saved_grad_shard = p.grad.data
+                    p.grad = None
+                else:
+                    assert False, "unexpected grad shape"
```
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1568]:
```suggestion
                    raise AssertionError(f"unexpected grad shape: {p.grad.size()}"
```

File: tests/nn/data_parallel/test_fsdp_grad_acc.py [hunk 4]
```diff
@@ -17,11 +18,15 @@
 from .test_fsdp import DistributedTest, NestedWrappedModule, rename_test, spawn_and_init
 
 
-class TestNoSync(DistributedTest):
+class TestGradAcc(DistributedTest):
```
Comment (sshleifer) [inline tests/nn/data_parallel/test_fsdp_grad_acc.py:21]: <3 naming

File: tests/nn/data_parallel/test_fsdp_grad_acc.py [hunk 5]
```diff
@@ -17,11 +18,15 @@
...
     def test_transformer(self):
         fn = functools.partial(self._test_transformer, config={})
         spawn_and_init(fn)
 
+    def test_transformer_grad_acc_without_no_sync(self):
+        fn = functools.partial(self._test_transformer, config={}, use_no_sync_context=False)
+        spawn_and_init(fn)
```
Comment (sshleifer) [inline tests/nn/data_parallel/test_fsdp_grad_acc.py:28]: would be cool to assert same loss between the two configs. Since the only tradeoff seems is speed vs. mem.

**Review by sshleifer — APPROVED** [review]
LGTM, really clear! 
@myleott ping me if you want me to benchmark this on the fairseq side.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 2 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:785]: I think this name comes from DDP, which has this no_sync context. I think we want to keep FSDP and DDP the same. I can certainly change it if people think otherwise.

File: tests/nn/data_parallel/test_fsdp_grad_acc.py (hunk 4 above)
Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp_grad_acc.py:21]: I changed the name since this is test both no_sync and without no_sync now. In both cases, it is testing grad acc. So I named it TestGradAcc. Let me know or just put in a github suggestion for a better name.

File: tests/nn/data_parallel/test_fsdp_grad_acc.py (hunk 5 above)
Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp_grad_acc.py:28]: Sure, let me look into it.

**Comment by myleott** [comment]
This is great, thanks for the quick fix! Tested in fairseq and it's working well. Is it possible to do a new release with this fix?

**Comment by min-xu-ai** [comment]
> 
> 
> This is great, thanks for the quick fix! Tested in fairseq and it's working well. Is it possible to do a new release with this fix?

Of course!
